operator/src/main/java/org/keycloak/operator/controllers/KeycloakAdminSecretDependentResource.java

```diff
@@ -0,0 +1,45 @@
+package org.keycloak.operator.controllers;
+
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.SecretBuilder;
+import io.fabric8.kubernetes.client.utils.KubernetesResourceUtil;
+import io.javaoperatorsdk.operator.api.reconciler.Context;
+import io.javaoperatorsdk.operator.api.reconciler.dependent.GarbageCollected;
+import io.javaoperatorsdk.operator.processing.dependent.Creator;
+import io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependent;
+import io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependentResource;
+
+import org.keycloak.operator.Constants;
+import org.keycloak.operator.Utils;
+import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
+
+import java.util.UUID;
+
+@KubernetesDependent(labelSelector = Constants.DEFAULT_LABELS_AS_STRING)
+public class KeycloakAdminSecretDependentResource extends KubernetesDependentResource<Secret, Keycloak> implements Creator<Secret, Keycloak>, GarbageCollected<Keycloak> {
+
+    public KeycloakAdminSecretDependentResource() {
+        super(Secret.class);
+    }
+
+    @Override
+    protected Secret desired(Keycloak primary, Context<Keycloak> context) {
+        return new SecretBuilder()
+                .withNewMetadata()
+                .withName(getName(primary))
+                .withResourceVersion("1") // ensures that we can't overwrite an existing
```

Comment:
> Adding a fake resourceVersion to the desired state ensures that we can't overwrite something that already exists https://github.com/keycloak/keycloak/pull/22044/files#diff-de536673442234fa231195f9a564525251ae83108dfbb4a62cf84017d23e5243R30 - that removes any possibilty of overwriting existing state.
> 
> The downside is that if someone were to create their own initial admin secret it must have the owner reference set to the keycloak instance - otherwise the create only logic will keep failing and complete block any additional reconciliation.

If I understand correctly, this should be handled already by the `Creator` interface, is that correct?

Reply:
> If I understand correctly, this should be handled already by the Creator interface, is that correct?

It's handled automatically in the way described - if something else creates an initial admin secret without an owner reference (or removes the owner reference from what we create) it will effectively block all further reconciliation of the Keycloak.  

On the otherhand if we leave the resourceVersion unpopulated, then the effect of the serverSideApply will be to recreate the secret data.  I was thinking that we want to avoid this.

Reply:
> So by adding resourceVersion we're basically addressing the case when someone created the Secret without a proper owner reference.

It's for addressing any possibility of overwriting an existing one.  If the existing one lacks the proper ownerReference it's the pathological case.

> Without resourceVersion, we'll just overwrite the Secret data.

If it does not have the owner reference, we'll just overwrite it.  If it does have the owner reference it's a matter of timing - if it's known to the informer before the reconciliation, then we'll just use what is already there.  If not, then we'll overwrite it.

> I'd argue we can be opinionated. If someone tries to mimic the Secret that's basically an internal Operator resource, we can overwrite without failing.

Okay seemed best to check as the prior impementation worked only by name - so regardless of whether there was an owner reference it would just use what was already there.

Here it sounds like you are are okay with the possibility of overwriting.

Reply:
> Okay seemed best to check as the prior impementation worked only by name - so regardless of whether there was an owner reference it would just use what was already there.
> 
> Here it sounds like you are are okay with the possibility of overwriting.

It's a good point that the old behaviour was different. But that IMHO doesn't mean it was correct. :)

I think it makes sense to stick to default dependent resources behaviour, i.e. overwrite the secret if not clearly owned by the Keycloak CR. We don't expect the users to create their own at the moment, this is not supported but we don't need to get into a failing state if there is an existing Secret. WDYT?

TBH, I'm not much in favor of adding the dummy resourceVersion as it feels too much like a workaround, and as I mentioned I don't see much (if any) benefit in that either.

Reply:
> I think it makes sense to stick to default dependent resources behaviour, i.e. overwrite the secret if not clearly owned by the Keycloak CR. We don't expect the users to create their own at the moment, this is not supported but we don't need to get into a failing state if there is an existing Secret. WDYT?

That's fine, my goal is just to avoid making any assumptions the intended behavior.  I'll remove the resourceVersion so that any existing secret, with or without a resourceVersion, will be overwriten by the current operator.  At worst this will mean that these new initial admin credentials would be picked up on the next restart.